### PR TITLE
Fix/Possible Breaking change: Contextualmenu not sizing properly. 

### PR DESCRIPTION
**Pull request by joschect**
Also use virtual elements for onclick events.

### CODE REVIEW

File: src/components/Callout/CalloutContent.tsx [hunk 1]
```diff
@@ -173,4 +183,48 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
...
+      top: bounds.top + spaceFromEdgeOfBounds,
+      left: bounds.left + spaceFromEdgeOfBounds,
+      right: bounds.right - spaceFromEdgeOfBounds,
+      bottom: bounds.bottom - spaceFromEdgeOfBounds,
+      width: bounds.width - spaceFromEdgeOfBounds * 2,
+      height: bounds.height - spaceFromEdgeOfBounds * 2
+    };
+  }
+
+  private get _Bounds(): IRect {
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: remove this, just use methods, it's way more obvious about what's happening

**Comment by dzearing** [comment]
1. increase height on items to 400px, open charms, make sure no stack overflow.
2. without increasing, make sure charms renders correctly

File: src/components/Callout/Callout.Props.ts [hunk 2]
```diff
@@ -69,4 +69,10 @@ export interface ICalloutProps extends React.Props<Callout|CalloutContent>, IPos
    * @returns True if focus was set, false if it was not.
    */
   setInitialFocus?: boolean;
+
+  /**
+   * The distance that the callout should maintain from the edge of the bounds.
+   * @default 10
```
Comment (dzearing) [inline src/components/Callout/Callout.Props.ts]: change to 8

File: src/components/Callout/CalloutContent.tsx [hunk 3]
```diff
@@ -8,15 +8,16 @@ import {
   css,
   elementContains
 } from '../../Utilities';
-import { getRelativePositions, IPositionInfo } from '../../utilities/positioning';
+import { getRelativePositions, IPositionInfo, IRect } from '../../utilities/positioning';
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: maybe use IRectangle from common

File: src/components/Callout/CalloutContent.tsx [hunk 4]
```diff
@@ -8,15 +8,16 @@ import {
...
 import { focusFirstChild } from '../../utilities/focus';
+import { assign } from '../../Utilities';
 import { Popup } from '../Popup/index';
 import { BaseComponent } from '../../common/BaseComponent';
 import './Callout.scss';
 
 const BEAK_ORIGIN_POSITION = { top: 0, left: 0 };
 const OFF_SCREEN_POSITION = { top: -9999, left: 0 };
-
+const borderWidth: number = 1;
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: BORDER_WIDTH?

File: src/components/Callout/Callout.Props.ts [hunk 5]
```diff
@@ -69,4 +69,10 @@ export interface ICalloutProps extends React.Props<Callout|CalloutContent>, IPos
...
+   */
+  spaceFromEdgeOfBounds?: number;
```
Comment (dzearing) [inline src/components/Callout/Callout.Props.ts]: i think this should move to constant only.

File: src/components/Callout/CalloutContent.tsx [hunk 6]
```diff
@@ -81,23 +85,26 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
       width: beakStyleWidth
     };
 
+    let contentMaxHeight: number = this._MaxHeight;
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: call method

File: src/components/Callout/CalloutContent.tsx [hunk 7]
```diff
@@ -81,23 +85,26 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
...
+
     let content = (
       <div ref={ this._resolveRef('_hostElement') } className={ 'ms-Callout-container' }>
         <div
-          className={ css('ms-Callout', className, slideDirectionalClassName ? `ms-u-${ slideDirectionalClassName }` : '') }
-          style={ ((positions) ? positions.callout : OFF_SCREEN_POSITION) }
+          className={ css('ms-Callout', className, slideDirectionalClassName ? `ms-u-${slideDirectionalClassName}` : '') }
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: break into multiple line

File: src/components/Callout/CalloutContent.tsx [hunk 8]
```diff
@@ -156,13 +171,21 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
...
 
       // Set the new position only when the positions are not exists or one of the new callout positions are different.
       // The position should not change if the position is within 2 decimal places.
       if ((!positions && positionInfo) ||
-        (positions && positionInfo && (positions.callout.top.toFixed(2) !== positionInfo.calloutPosition.top.toFixed(2) || positions.callout.left.toFixed(2) !== positionInfo.calloutPosition.left.toFixed(2)))) {
-
+        (positions && positionInfo &&
+          (positions.callout.top.toFixed(2) !== positionInfo.calloutPosition.top.toFixed(2) ||
+            positions.callout.left.toFixed(2) !== positionInfo.calloutPosition.left.toFixed(2))
+          && this._positionAttempts < 10000)) {
```
Comment (dzearing) [inline src/components/Callout/CalloutContent.tsx]: hmmm this looks like a gigantic number of allowed attempts.

File: src/components/ContextualMenu/ContextualMenu.Props.ts [hunk 9]
```diff
@@ -81,6 +81,10 @@ export interface IContextualMenuProps extends React.Props<ContextualMenu>, IPosi
    * If none specified no aria label will be applied to the ContextualMenu.
    */
   ariaLabel?: string;
+  /**
```
Comment (dzearing) [inline src/components/ContextualMenu/ContextualMenu.Props.ts]: nit space between props

File: src/components/Callout/CalloutContent.tsx (hunk 8 above)
Comment (joschect) [inline src/components/Callout/CalloutContent.tsx]: Whoops! I never changed that back lol.

File: src/components/Callout/CalloutContent.tsx [hunk 10]
```diff
@@ -171,6 +194,45 @@ export class CalloutContent extends BaseComponent<ICalloutProps, ICalloutState>
...
+          width: this._targetWindow.innerWidth,
+          height: this._targetWindow.innerHeight
+        };
+      }
+      this._bounds = this._modifyBoundsForPadding(currentBounds);
+    }
+    return this._bounds;
+  }
+
+  private _modifyBoundsForPadding(bounds: IRectangle): IRectangle {
```
Comment (joschect) [inline src/components/Callout/CalloutContent.tsx]: Yes it can be. I just like lots of methods I guess in this case it's really not necessary.

**Pull request merged**